dev/src/bulk-writer.ts

```diff
@@ -708,29 +867,44 @@ export class BulkWriter {
    * @private
    */
   private sendReadyBatches(): void {
-    const unsentBatches = this.batchQueue.filter(
-      batch => batch.state === BatchState.READY_TO_SEND
-    );
-
-    let index = 0;
-    while (
-      index < unsentBatches.length &&
-      unsentBatches[index].state === BatchState.READY_TO_SEND
-    ) {
+    // Always send retry batches first if the retryBatchQueue isn't empty.
+    const sendRetries = this._retryBatchQueue.length > 0;
+    const unsentBatches = sendRetries
+      ? this._retryBatchQueue
+      : this._batchQueue;
+    for (let index = 0; index < unsentBatches.length; ++index) {
       const batch = unsentBatches[index];
 
+      // Send all retries whenever possible.
+      if (sendRetries) {
+        batch.markReadyToSend();
```

Comment:
I was wondering if we need this call? Could we not transition from OPEN to SEND? If we do need it (I haven't thought it quite through) then maybe it makes sense to post-process the retry queue before entering here.